lib/rhykane/functions.rb

```diff
@@ -1,15 +1,56 @@
 # frozen_string_literal: true
 
+require 'active_support/core_ext/date/calculations'
+
 module Functions
   def to_json(value)
     JSON.dump(value) unless value.nil?
   end
 
-  def parse_period(value, type)
+  def parse_period(value, args)
     return if value.nil?
 
-    date = Date.strptime(value.slice(2..-1), '%m%y')
-    day = type == :start ? 1 : -1
-    Date.new(date.year, date.month, day).to_s
+    date, day = get_day(value, **args)
+    return create_date(date, day) unless args[:quarter_type]
+
+    qtr_args = { type: args[:type], date:, day: }
+    args[:quarter_type] == :ordinal ? ordinal_qtr(**qtr_args) : numeric_qtr(**qtr_args)
+  end
+
+  def get_day(value, date_format:, type:, **)
+    date = Date.strptime(value, date_format)
+    day  = type == :start ? 1 : -1
+    [date, day]
+  end
+
+  def create_date(date, day, month = nil)
+    month ||= date.month
+    Date.new(date.year, month, day).to_s
+  end
+
+  def ordinal_qtr(type:, date:, day:)
+    month = type == :start ? (date.month * 3) - 2 : date.month * 3
+    create_date(date, day, month)
+  end
+
+  def numeric_qtr(type:, date:, day:)
+    month = type == :start ? date.month - 2 : date.month
+    create_date(date, day, month)
+  end
+
+  def seconds_to_iso(original)
+    duration = ActiveSupport::Duration.build(original.to_f.round)
+    h = duration.parts[:hours]   || 0
```

Comment:
Can we use the `#iso8601` method instead?

Reply:
The `#iso8601` will omit the hour, minute, or second fields if they're equal to zero, and we want to include them.
For example, here's what we want to avoid:
```ruby
duration = ActiveSupport::Duration.build(60)
duration.iso8601 # returns "PT1M"
```
Instead, we want this:
```ruby
    duration = ActiveSupport::Duration.build(60)
    h = duration.parts[:hours]   || 0
    m = duration.parts[:minutes] || 0
    s = duration.parts[:seconds] || 0

    "PT#{h}H#{m}M#{s}S" # returns "PT0H1M0S"
```

Reply:
Thanks, refactored